### ISSUE TITLE
Hello, question for VBA code to get Card data form BEID after march 2014

### ISSUE THREAD

**Issue by JohanRVT** [jdt/EID.Wrapper#2]
Hi Jesper nice work with that wrapper. I installed everything on our Windows 64, MS Access 2007 and it works nicely with this code to get a messagebox with some of the card data:

Private Sub CmdEIDTest_Click()
Dim wrapper As New EID_Wrapper.wrapper
Dim data As EID_Wrapper.CardData
         Set data = wrapper.GetCardData()
' Toon de EID Kaart Data
MsgBox "Voornamen: " & (data.FirstNames) & vbCrLf & _
       "Naam: " & (data.Surname) & vbCrLf & _
       "Geslacht: " & (data.Gender) & vbCrLf & _
       "Geboortedatum: " & (data.BirthDate) & vbCrLf & _
       "Geboorteplaats: " & (data.BirthPlace) & vbCrLf & _
       "Straat en nummer: " & (data.StreetAndNumber) & vbCrLf & _
       "Postcode: " & (data.ZipCode) & vbCrLf & _
       "Woonplaats: " & (data.Municipality) & vbCrLf & _
       "Nationaliteit: " & (data.Nationality) & vbCrLf & _
       "Rijksregisternummer: " & (data.NationalNumber), vbInformation + vbOKOnly, "EID Kaart info"
End Sub

'Old Fields that don't work with this technique
'"Kaartnummer: " & (data.CardNumber) & vbCrLf & _'
'"Chipnummer: " & (data.ChipNumber) & vbCrLf & _'
'"Uitgifteplaats: " & (data.IssuingMunicipality) & vbCrLf & _'
'"Beginvaliditeit: " & (data.BeginValidityDate) & vbCrLf & _'
'"Eindvaliditeit: " & (data.EndValidityDate) & vbCrLf & _'

Below here is the code that I made to get the data to a form but I can't get ALL the data and the picture to the form

Private Sub CmdEIDTest2_Click()
On Error GoTo Err_CmdEIDTest2_Click
'Uitvoerende VBA code
Dim wrapper As New EID_Wrapper.wrapper
   Dim data As EID_Wrapper.CardData
   'BEID Persoons Info
   Dim strName As String
   Dim dirname As String
   Dim strFirstName1 As String
   Dim strConvertedFirstname As String
   Dim strConvertedSecondFirstname As String
   Dim strBirthPlace As String
   Dim strBirthDate As String
   Dim strConvertedBirthDate As String
   Dim strGender As String
   Dim strconvertedgender As String
   Dim strNationality As String
   Dim strNationalNumber As String

   'BEID Adres Info
   Dim notexists As Integer
   Dim exists As Integer
   Dim strstreetNumber As String
   Dim strConvertedStreet As String
   Dim strConvertedStreetNew As String
   Dim strZipCode As String
   Dim strMunicipality As String
   Dim iPos As Integer
   Dim iPos2 As Integer

```
Set data = wrapper.GetCardData()
' Lees de EID Kaart Data
strFirstName1 = (data.FirstNames)
'methode indien de originele voornaam uit 2 woorden bestaat die dan omzettten naar twee aparte voornamen.
' oude identiteitskaarten had (soms) meerdere voornamen; de nieuwe slechts één
    If exists = InStr(strFirstName1, " ") = 0 Then
          strConvertedFirstname = Left([strFirstName1], InStr(1, [strFirstName1], " ") - 1)
          strConvertedSecondFirstname = Right([strFirstName1], Len([strFirstName1]) - InStrRev([strFirstName1], " "))
    Else
          strConvertedFirstname = (data.FirstNames)
    End If
strName = (data.Surname)
strGender = (data.Gender)
            If notexists = InStr(strGender, "M") = 0 Then
                strconvertedgender = (data.Gender)
            End If
            If exists = InStr(strGender, "M") <> 0 Then
                strconvertedgender = "V"
             End If
strNationalNumber = (data.NationalNumber)
strBirthDate = (data.BirthDate) ' de maand wordt in letters weergegeven en dan nog inconsistent; niet bruikbaar.
'methode omzetten van rijksregisternummer naar geboortedatum; veel preciezer dus deze gebruiken
strConvertedBirthDate = CDate(Mid$(strNationalNumber, 5, 2) & "/" & Mid$(strNationalNumber, 3, 2) & "/19" & Left$(strNationalNumber, 2))
strBirthPlace = (data.BirthPlace)
strZipCode = (data.ZipCode)
strMunicipality = (data.Municipality)
strNationality = (data.Nationality)
strstreetNumber = (data.StreetAndNumber)
' google maps werkt niet goed met een straatnaam met gemeenteafkorting tussen haakjes in
' indien het zo op de EID staat geeft dit de straat weer zonder dat er haakjes in staan met de deelgemeente

        If exists = InStr(strstreetNumber, "(") <> 0 Or InStr(strstreetNumber, ")") <> 0 Then
                    iPos = InStr(strstreetNumber, "(")
                    iPos2 = InStr(strstreetNumber, ")")
                    If iPos > 0 Then
                        strConvertedStreet = Left$(strstreetNumber, iPos - 1) & "" & Mid$(strstreetNumber, iPos2 + 1)
                     End If
        End If
'velden invullen op het formulier met de data die we momenteel uit de BEID kaart kunnen halen met de wrapper

          Me.BNaam = Replace(Replace(Replace(strName, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è")
          Me.BVoornaam = Replace(Replace(Replace(strConvertedFirstname, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è")
          Me.[B2e voornaam] = Replace(Replace(Replace(strConvertedSecondFirstname, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è")
          Me.BGeboorteplaats = strBirthPlace
          Me.BGeboortedatum = strConvertedBirthDate
          Me.TxtNationaliteit = strNationality
          Me.BRijksregisternummer = strNationalNumber
          Me.[BVPostcode vorige woonpl] = strZipCode
          Me.TxtBHuidige_postcode = strZipCode
          Me.BVWoonplaats = Replace(Replace(Replace(strMunicipality, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è")
          Me.TxtBHuidige_Woonplaats = Replace(Replace(Replace(strMunicipality, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è")
          Me.[BVStraat vorige woonpl] = strConvertedStreet
          Me.TxtBHuidige_straat = strConvertedStreet
          Me.[m/v] = strconvertedgender
          Me.TxtTotaal.Requery
          Me.AllowAdditions = False
          dirname = GetPath & "\Bewonersdossier\Fiches\" & Me.BNaam & " " & Me.BVoornaam
          If Dir(dirname, vbDirectory) = "" Then MkDir dirname
```

Exit_CmdEIDTest2_Click:
     Exit Sub
Err_CmdEIDTest2_Click:
     Select Case Err.Number
            Case 20 'Resume zonder Error
                    Resume Next
            Case 94 'Ongeldig gebruik van Null
                 MsgBox " Geen criteria, herbegin of vul het nodige veld in", vbCritical + vbOKOnly, "Opgelet!"
                    Resume Next
            Case 3021 ' Geen huidige record
                    Resume Next
            Case 3077  'Geen criteria
                    MsgBox " Geen criteria, herbegin", vbCritical + vbOKOnly, "Opgelet!"
                    Resume Next
            Case 3167  'Record is al verwijderd
                    Resume Next
            Case 9999
                   Resume Next
            Case 999
                   Resume Exit_CmdEIDTest2_Click
            Case Else
            Call LogError(Err.Number, Err.Description, "CmdEIDTest2_Click()")
                Resume Exit_CmdEIDTest2_Click
     End Select
End Sub

Below is my old VBA code (not complete but works fine with the old BEID's; how can I transform that to do the same with the new BEID's

Private Sub cmdEID_Click()
 On Error GoTo Err_cmdEID_Click
   Dim EIDlib1              As New EIDLIBCTRLLib.EIDlib
   Dim lhandle              As Long
   Dim RetStatus            As New EIDLIBCTRLLib.RetStatus
   Dim MapColPicture        As New EIDLIBCTRLLib.MapCollection
   Dim MapColID             As New EIDLIBCTRLLib.MapCollection
   Dim MapColAddress        As New EIDLIBCTRLLib.MapCollection
   Dim CertifCheck          As New EIDLIBCTRLLib.CertifCheck

'BEID Persoons Info
   Dim strName As String
   Dim dirname As String
   Dim strFirstName1 As String
   Dim strConvertedFirstname As String
   Dim strConvertedSecondFirstname As String
   Dim strBirthPlace As String
   Dim strBirthDate As String
   Dim strConvertedBirthDate As String
   Dim strConvertedNewBirthDate As String
   Dim strGender As String
   Dim strconvertedgender As String
   Dim strNationality As String
   Dim strNationalNumber As String

'BEID Kaart Info
   Dim strCardNumber As String
   Dim strChipNumber As String
   Dim strConvertedBegVal As String
   Dim strBegValDag As String
   Dim strBegVal As String
   Dim strBegValMaand As String
   Dim strBegValJaar As String
   Dim strConvertedEndVal As String
   Dim strEndValDag As String
   Dim strEndVal As String
   Dim strEndValMaand As String
   Dim strconvertedmonth As String
   Dim strconvertedday As String
   Dim strIssuingMunicipality As String
   Dim strdate As String
   Dim dteConvertedDate As Date
'BEID pasfoto Info
   Dim Pasfoto_Temp() As Byte
   Dim filename As String
   Dim strFileName As String
   Dim iPos As Integer
   Dim iPos2 As Integer
'BEID Adres Info
   Dim notexists As Integer
   Dim exists As Integer
   Dim strStreet As String
   Dim strConvertedStreet As String
   Dim strConvertedStreetNew As String
   Dim strZipCode As String
   Dim strMunicipality As String

' Initialisatie van de EID, loopt mis met de EID's uitgereikt na maart 2014

   Set RetStatus = EIDlib1.Init("", 0, 0, lhandle)
   Set RetStatus = EIDlib1.GetID(MapColID, CertifCheck)
' Nakijken of er al een kaart in de reader is gestoken en indien niet een bericht geven
If Not (RetStatus.GetGeneral = 0) Then
        MsgBox "Geen EId kaart ingevoegd! " & vbCrLf & _
        "Steek een Identiteitskaart in de kaartlezer, " & vbCrLf & _
        "of controleer of ze goed is geplaatst!", vbCritical + vbOKOnly, "Opgelet!"
            Exit Sub
End If
' Nakijken of er al een naam is ingevuld
Select Case IsNull(Me.BNaam.Value)
Case Is = 0
        Dim LResponse As Integer
            LResponse = MsgBox("Er werd al een naam " & BNaam & " " & BVoornaam & " ingevuld," & vbCrLf & _
            "wilt u de gegevens overschrijven ?", vbCritical + vbYesNo, "Opgelet!")

```
If LResponse = vbYes Then


       strName = MapColID.GetValue("Name")
       strFirstName1 = MapColID.GetValue("FirstName1")
          'methode indien de originele voornaam uit 2 woorden bestaat die dan omzettten naar twee aparte voornamen.
          ' oude identiteitskaarten had (soms) meerdere voornamen; de nieuwe slechts één
          If exists = InStr(strFirstName1, " ") = 0 Then
                strConvertedFirstname = Left([strFirstName1], InStr(1, [strFirstName1], " ") - 1)
                strConvertedSecondFirstname = Right([strFirstName1], Len([strFirstName1]) - InStrRev([strFirstName1], " "))
           Else
                strConvertedFirstname = MapColID.GetValue("FirstName1")
          End If
       strBirthPlace = MapColID.GetValue("BirthPlace")
      'methode  omzettten van origineel ingescande geboortedatum; loopt soms nog mis dus niet meer gebruikt
       strBirthDate = MapColID.GetValue("BirthDate")
            If CDate(Mid$(strBirthDate, 5, 2)) > CDate(Right$(strBirthDate, 2)) Then
                strConvertedBirthDate = CDate(Right$(strBirthDate, 2) & "/" & Mid$(strBirthDate, 5, 2) & "/" & Left$(strBirthDate, 4))
            Else
                strConvertedBirthDate = CDate(Mid$(strBirthDate, 5, 2) & "/" & Right$(strBirthDate, 2) & "/" & Left$(strBirthDate, 4))
            End If
       strGender = MapColID.GetValue("Gender")
            If notexists = InStr(strGender, "M") = 0 Then
                strconvertedgender = MapColID.GetValue("Gender")
            End If
            If exists = InStr(strGender, "M") <> 0 Then
                strconvertedgender = "V"
             End If
       strNationality = MapColID.GetValue("Nationality")
       strNationalNumber = MapColID.GetValue("NationalNumber")
       'methode omzetten van rijksregisternummer naar geboortedatum; veel preciezer dus deze gebruiken
       strConvertedNewBirthDate = CDate(Mid$(strNationalNumber, 5, 2) & "/" & Mid$(strNationalNumber, 3, 2) & "/19" & Left$(strNationalNumber, 2))
       Set RetStatus = EIDlib1.GetAddress(MapColAddress, CertifCheck)
       strZipCode = MapColAddress.GetValue("ZIPCode")
       strMunicipality = MapColAddress.GetValue("Municipality")
       strCardNumber = MapColID.GetValue("CardNumber")
       strChipNumber = MapColID.GetValue("ChipNumber")
       strIssuingMunicipality = MapColID.GetValue("IssuingMunicipality")
       strBegVal = MapColID.GetValue("BeginValidityDate")
       strConvertedBegVal = CDate(Right$(strBegVal, 2) & "/" & Mid$(strBegVal, 5, 2) & "/" & Left$(strBegVal, 4))
       strEndVal = MapColID.GetValue("EndValidityDate")
       strConvertedEndVal = CDate(Right$(strEndVal, 2) & "/" & Mid$(strEndVal, 5, 2) & "/" & Left$(strEndVal, 4))

       'Haal eID foto op
       Set RetStatus = EIDlib1.GetPicture(MapColPicture, CertifCheck)
       Pasfoto_Temp = MapColPicture.GetValue("Picture")
       'Schrijf eID foto weg naar een bestand
       filename = GetPath & "\fotomap\" & Replace(Replace(Replace(strName, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è") & "_" & Replace(Replace(Replace(strConvertedFirstname, "Ã«", "ë"), "Ã©", "é"), "Ã¨", "è") & "_eid.jpg"
       SavePicture PictureFromRes(Pasfoto_Temp), filename

       'Laad bestand in image control
       ImageFrame.Picture = filename
        strStreet = MapColAddress.GetValue("Street")
       ' google maps werkt niet goed met een straatnaam met gemeenteafkorting tussen haakjes in
       ' indien het zo op de EID staat geeft dit de straat weer zonder dat er haakjes in staan met de deelgemeente

        If exists = InStr(strStreet, "(") <> 0 Or InStr(strStreet, ")") <> 0 Then
                    iPos = InStr(strStreet, "(")
                    iPos2 = InStr(strStreet, ")")
                    If iPos > 0 Then
                        strConvertedStreet = Left$(strStreet, iPos - 1) & "" & Mid$(strStreet, iPos2 + 1)
                     End If
        End If

       Set RetStatus = EIDlib1.Exit
```

I am new to this site be need some help.

Thanks in advance

Also available for PM at kine.marialove at olv7w.org

**Comment by jdt** [comment]
Hello Johan, thank you for the feedback. I've added the additional fields you need (card- and chipnumber, issuing municipality and validity begin- and enddates) as well as the ability to read and work with the passport photograph on the card. I've updated the README.md with the needed information to get you going so you should download the new files and check it out!

**Comment by JohanRVT** [comment]
Hello Jesper, Thanks for the update. Now it's working for card- and chipnumber, issuing municipality and validity begin- and enddates. Also the picture import is nice.  With the console.exe it is also working and can I see the picture that is saved to my tmp directory (see picture below). I see that the picture now is only 6Kb large in stead of the usual 83Kb;
I also noticed that the cardstatus takes about 8 seconds to complete where the old check took about 4 seconds.
![tmp7ea4 tmp](https://cloud.githubusercontent.com/assets/11335415/6573922/e4c0095c-c71f-11e4-9dc0-fbb346bf3fe3.jpg)

**Comment by jdt** [comment]
I wasn't really sure from your previous message, are you still having issues with this library?

**Comment by JohanRVT** [comment]
Jesper,
Your updated wrapper worked fine on our 2X  netwerk but it doesn’t  work for the moment on the new Citrix server based connected clients. Wrapper is installed on the server.
The speed of reading out a EID was 1’25” where the old beidlib software did it in 15”

Van: Jesper De Temmerman [mailto:notifications@github.com]
Verzonden: woensdag 8 april 2015 20:59
Aan: jdt/EID.Wrapper
CC: Marialove - Kiné
Onderwerp: Re: [EID.Wrapper] Hello, question for VBA code to get Card data form BEID after march 2014 (#2)

I wasn't really sure from your previous message, are you still having issues with this library?

**Comment by JohanRVT** [comment]
Hello Jesper,

De wrapper is now working on Citrix.

Van: Jesper De Temmerman [mailto:notifications@github.com]
Verzonden: woensdag 8 april 2015 20:59
Aan: jdt/EID.Wrapper
CC: Marialove - Kiné
Onderwerp: Re: [EID.Wrapper] Hello, question for VBA code to get Card data form BEID after march 2014 (#2)

I wasn't really sure from your previous message, are you still having issues with this library?

**Comment by didierr** [comment]
Hi,

I have the same problem as JohanRVT.
The code is working well for eid issued before march 2014 but after march 2014, the code is not wording.

"jdt" was speaking about an update in a README file but I did'nt see this file.

As someone a solution for reading belgian eid issued after march 2014 ?

Best regards,

DidierR

**Comment by gerpat** [comment]
Hello,
Thanks a lot for your work and cooperation.
I have the same problem as JohanRVT.
I am working on a PC with windows 10 (64) and office 2016 (64).
Is not possible to get card- and chip number, issuing municipality and validity begin- and end dates and the picture.
I have downloaded your last EID.Wrapper-master and followed the instructions of README.md but it's still not working. The data.FisrtCard ist not available.
This is a big issue for me because I am in charge of the database of my club (more than 400 members). I have made a C# application that can read the EiD data and then export to the Access DB, but is not a proper way to work.
Many thanks in advance for your answer.
